internal/resolver/resolver.go

```diff
@@ -1144,33 +1147,106 @@ func (r *resolver) loadNodeModules(path string, kind ast.ImportKind, dirInfo *di
 	for _, absDir := range r.options.AbsNodePaths {
 		absPath := r.fs.Join(absDir, path)
 		if absolute, ok, diffCase := r.loadAsFileOrDirectory(absPath, kind); ok {
-			return absolute, true, diffCase
+			return absolute, true, diffCase, nil
 		}
 	}
 
+	esmPackageName, esmPackageSubpath, esmOK := esmParsePackageName(path)
+
 	// Then check for the package in any enclosing "node_modules" directories
 	for {
 		// Skip directories that are themselves called "node_modules", since we
 		// don't ever want to search for "node_modules/node_modules"
 		if dirInfo.hasNodeModules {
 			absPath := r.fs.Join(dirInfo.absPath, "node_modules", path)
 
+			// Check for an "exports" map in the package's package.json folder
+			if esmOK {
+				absPkgPath := r.fs.Join(dirInfo.absPath, "node_modules", esmPackageName)
+				if pkgDirInfo := r.dirInfoCached(absPkgPath); pkgDirInfo != nil {
+					if pkgJSON := pkgDirInfo.packageJSON; pkgJSON != nil && pkgJSON.exportsMap != nil {
+						// Resolve against the path "/", then join it with the absolute
+						// directory path. This is done because ESM package resolution uses
+						// URLs while our path resolution uses file system paths. We don't
+						// want problems due to Windows paths, which are very unlike URL
+						// paths. We also want to avoid any "%" characters in the absolute
+						// directory path accidentally being interpreted as URL escapes.
+						resolvedPath, status, token := esmPackageExportsResolveWithPostConditions("/", esmPackageSubpath, pkgJSON.exportsMap.root, nil)
```

Comment:
Where do the `conditions` come from? AFAICT this is the only initialization point for the `exports` resolution phase

Reply:
They were added in the very next commit. The first commit only respects the `default` condition and then the next commit adds `browser`, `node`, `import`, and `require`.

Reply:
Oh, haha 😅 thanks I'll look now

Reply:
Looks great! The only thing I'm not seeing – i think – is support for custom conditions? 

For example, webpack adds `development` -vs- `production` conditions & a lot of people are making use of that already. I've also seen (and have authored) modules with a `worker` condition. Technically anything can be a condition, and so long as resolver it willing to _look_ at it & the pkg author has defined it in the correct order-hierarchy, then it should be matched.

Reply:
I'm already aware of the need for conditional exports. I'm still working on the 0.9.0 release and just haven't implemented them yet.